### DISCUSSION TITLE
Livewire full page component not working with "can" middleware

### DISCUSSION THREAD

**Discussion by boptom** [livewire/livewire#8445] [7. Bugs (archived)]
### Livewire version

v3.4.12

### Laravel version

v11.7

### Which PHP version are you using?

PHP 8.3

### Steps To Reproduce

```
git clone https://github.com/boptom/volt-route-middleware
cd volt-route-middleware
composer install

cp .env.example .env
php artisan key:generate
php artisan migrate

php artisan test tests/Feature/PostTest.php
```

### Repository or Wirebox URL

https://github.com/boptom/volt-route-middleware

### Additional information

Using 'can:view,post' in middleware in Volt route or Livewire full page component always returns 403.

```
Volt::route('post/{post}', 'post')
    ->middleware(['auth', 'can:view,post'])
    ->name('post');

Route::get('livewire-post/{post}', LivewirePost::class)
    ->middleware(['auth', 'can:view,post'])
    ->name('livewire-post');
```

Previously reported in the Volt repo here: https://github.com/livewire/volt/issues/104

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by PerryvanderMeer** [comment]
You should type-hint the `$post` property in your component like `public Post $post` .
See https://livewire.laravel.com/docs/components#using-route-model-binding and https://livewire.laravel.com/docs/security#middleware

Not sure about Volt, that may be a limitation in Volt's functionality.

**Reply by boptom** [reply]
Hey thanks!

That seems to have fixed the livewire full page component.

However, Volt seems to still be broken. Here is the updated Volt page with the Post type:

https://github.com/boptom/volt-route-middleware/blob/main/resources/views/livewire/post.blade.php

```
<?php

use Livewire\Attributes\Layout;
use App\Models\Post;
use Livewire\Volt\Component;

new
#[Layout('layouts.app')]
class extends Component
{
    public Post $post;

    public function mount(Post $post)
    {
        $this->post = $post;
    }
}

?>

<div>
    Post View {{ $post->id }}
</div>
```

**Comment by IamSAB** [comment]
Has this issue been fixed? Just ran into it now in Volt full-page components.

```php
// web.php
Volt::route('contact/{contact}', 'pages.admin.contact')
            ->name('contact.edit')
            ->can('update', 'contact');

// blade file
new class extends Component
{
    public Contact $contact;

    public function mount(Contact $contact)
    {
        $this->contact = $contact;
    }
}
```
Policy does not get called (as I do not get logs from the before() hook).

```php
class ContactPolicy
{
    public function before(Contact $user, string $ability): bool|null
    {
        Log::debug('was here');

        return null;
    }
}
```

**Reply by boptom** [reply]
I don’t think it’s been fixed. I’m trying to re-open an issue in the Volt repo that was closed. 

https://github.com/livewire/volt/issues/104

Maybe comment there too for visibility?

**Reply by firer777** [reply]
you must use Route::model('post', App\Models\Post::class); on Providers/AppServiceProvider.php for resolve post in livewire

**Comment by IamSAB** [comment]
How to resolve this issue?

I used the ` Route::model('post', Post::class)` approach but then i cannot define custom keys. I want to use in the frontend slugs and in the backend ids.

```php

Volt::route('post/{post:slug}', 'pages.site.post')
    ->name('post');
...
Route::middleware('auth')->group(function () {
   ...
    Route::prefix('admin')->group(function () {
       Volt::route('post/{post}', 'pages.admin.post')
            ->name('posts.edit')
            ->can('update', 'post');
   }
}
```
In this case, explicitely binding, the slug route throws 404, when allow implicit binding the id route throws a 403.

I there a way to manually bind the model per route?

**Reply by firer777** [reply]
Route::bind('post', function(mixed $value, $route) {
            return Post::where($route->bindingFieldFor('post') ?: 'id', $value)->firstOrFail();
});

**Reply by firer777** [reply]
If u want with trashed u cant wire like this:

Route::model('post',  function(mixed $value, $route) {
       $withTrashed = $route->allowsTrashedBindings();
      $query =  Post::where($route->bindingFieldFor('post') ?: 'id', $value);
     if ($withTrashed) {
           $query->withTrashed();
     }
     return $query->firstOrFail()
});

**Reply by IamSAB** [reply]
Thanks a lot, works :)
